src/mscorlib/src/System/IO/MemoryStream.cs

```diff
@@ -402,31 +402,42 @@ public override Task<int> ReadAsync(Byte[] buffer, int offset, int count, Cancel
             return _buffer[_position++];
         }
 
+        public override void CopyTo(Stream destination, int bufferSize)
+        {
+            // Since we did not originally override this method, validate the arguments
+            // the same way Stream does for back-compat.
+            ValidateCopyToArguments(destination, bufferSize);
 
-        public override Task CopyToAsync(Stream destination, Int32 bufferSize, CancellationToken cancellationToken) {
-
-            // This implementation offers beter performance compared to the base class version.
-
-            // The parameter checks must be in sync with the base version:
-            if (destination == null)
-                throw new ArgumentNullException("destination");
+            // If we have been inherited into a subclass, the following implementation could be incorrect
+            // since it does not call through to Read() or Write() which a subclass might have overriden.  
+            // To be safe we will only use this implementation in cases where we know it is safe to do so,
+            // and delegate to our base class (which will call into Read/Write) when we are not sure.
+            if (GetType() != typeof(MemoryStream))
+            {
+                base.CopyTo(destination, bufferSize);
+                return;
+            }
             
-            if (bufferSize <= 0)
-                throw new ArgumentOutOfRangeException("bufferSize", Environment.GetResourceString("ArgumentOutOfRange_NeedPosNum"));
+            int originalPosition = _position;
+            int remaining = _length - originalPosition;
```

Comment:
IMO this is clearer than `InternalEmulateRead`. Also, we already know certain invariants at this point (`_isOpen` is true, `n == count`) so this will be faster, plus calling that method isn't likely to be inlined. I'm thinking of changing `CopyToAsync` to not use that method in a later PR, so we can delete it altogether.